### ISSUE TITLE
design for comprehensive styling in latex

### ISSUE THREAD

**Issue by brainchild0** [jgm/pandoc#6394]
# Problem

## Overview

LaTeX is among the most popular target formats in current use with Pandoc. Similar to languages such as Markdown and HTML, LaTeX is a semantic language. Utilizing its full capabilities and flexibility depends on expressing a document structure through semantic demarcation, applied to rules for visual formatting style defined separately. 

Currently, Pandoc inconsistently employs semantic demarcation of abstract types, that is, Pandoc types, in generated LaTeX documents. To represent an abstract document type that corresponds to a core feature of LaTeX, Pandoc utilizes that feature. For every other type, Pandoc instead generates a sequence of macros that directly determine visual formatting.

The inconsistency produces a challenge against the user accessibly and flexibly adding custom styling to the full breadth of textual features throughout a document.

## Example

Consider the following short document in Markdown:

```markdown
Do you *strongly* agree?

---

I **very strongly** agree.
```

Pandoc generates the following result when converting to LaTeX:

```latex
Do you \emph{strongly} agree?

\begin{center}\rule{0.5\linewidth}{0.5pt}\end{center}

I \textbf{very strongly} agree.
```

The document includes several textual demarcations.

The first demarcation, basic emphasis, is represented in the result using the macro `\emph`, whose purpose is to apply appropriate visual styling to text carrying precisely the semantic distinction of basic emphasis.

The further demarcations, a thematic break and strong emphasis, are not converted analogously, but rather represented by macros directly determining visual formatting.

## Analysis

In the case of basic emphasis, the `\emph{}` macro, in contrast to `\textit{}`, though many casually view it as essentially the same, enables specification of arbitrary visual style. In any particular case, the preferred style may be the default, which is generally an italic font. Otherwise, it may be a very similar style, such as an oblique font, or it may be rather a very different one. Further, `\emph{}` has the particular behavior of redefining itself within its own scope, such as to produce a different effect in a nested context. Generally, text within a doubly-nested invocation of the macro has the same appearance as text entirely outside of any invocation. Even this effect, though, may depend on context, as the macro may be redefined inside certain environments or other macros. The `\textit{}` macro lacks such context-dependent behavior, and adding it through customization would have detrimental effects.

Despite the important function for basic emphasis of `\emph{}`, many find adequate the conversion of strong emphasis through the `\textbf{}` macro. Because in many uses strong emphasis appears in input precisely to express the intention of heavier font weight, any other representation may seem unnecessary to consider. Yet even such a simple case carries a variety of weaknesses. Many font families provide a multiplicity of font weights, beyond the normal and bold weights familiar for casual formatting, with one or another often being preferred aesthetically depending on a particular context. Direct use of a macro, perhaps named `\strong{}`, to demarcate strong emphasis, instead of `\textbf{}`, would offer control over such details, by permitting precise customization of visual style according to context, while still retaining the benefit of an expected default behavior. Introduction of such a macro would complete a symmetry between the representation of strongly-emphasized text and of basic emphasized text.

Whereas some may experience no restriction from existing functionality, its limitations constrain users generally to a narrow band within the full breadth of cases and concerns.

# Objective

Users generating LaTeX documents in Pandoc currently have one experience with a particular class of abstract types, which includes basic emphasis, and another experience with a different class of abstract types, which includes strong emphasis. The reason for the difference is that elements of the latter class have no semantic representation in LaTeX.

Yet both classes have equal value and utility among Pandoc abstract document types.

A proposed enhancement eliminates this asymmetry.

The enhancement has the following features:

- The usage is *familiar*, expressing a clear and simple symmetry between handling of abstract types with support from core LaTeX, versus handling of other abstract types.
- The effects are *immediate*, available through basic invocation of the application, without any nontrivial customization or external integration.
- The behavior is *configurable*, supporting a qualified variety of particular and nonstandard use cases.
- The design is *robust*, depending only on supported and standard utilization of languages and tools.
- The functionality is *compatible*, preserving active and recurring use cases, and ensuring the ability to emulate previous operation.

The virtues of these features are plain. Common handling of the full breadth of semantic demarcation makes styling equally accessible in all cases through straightforward methods common to all users. Inclusion of the functionality directly into the application obviates any need of users to create a custom solution or to rely on a third-party one. Configuration options enable control over application behavior, for cases of particular requirements. Finally, while failure or volatility is unlikely, as is damage to current patterns of usage, the possibility to disable the feature offers a final protection against genuine harm.

# Solution

At the core of the proposed enhancement is a collection of LaTeX macros and environments, augmenting those native to the language, or defined in the standard classes and common packages, to complete the mapping between types in the abstract document model versus macros and environments in generated LaTeX documents.

After such an enhancement, the generated LaTeX fragment corresponding to the earlier example might appear instead as follows:

```latex
Do you \emph{strongly} agree?

\thematicbreak

I \strong{very strongly} agree.
```

Just as LaTeX provides a definition of `\emph`, the enhancement would provide definitions of `\thematicbreak` and `\strong`, offering the user in any context the possibility either simply to enjoy the default definitions, or, as preferred, to modify or to replace them.

## Modes

A basic design simply would expand the current document header to include definitions for additional macros and environments, whose invocations would appear where appropriate in the document body. The need for a solution that is configurable and compatible, however, demands a more sophisticated design. The proposed solution offers users a choice between three operational modes, called *full mode*, *in-line mode*, and *verbatim mode*.

### Full Mode

Under *full mode*, a generated LaTeX document invokes the additional macros and environments, where appropriate, instead of directly invoking visual formatting macros and environments, in the document body.

In the case of a standalone document, the header includes a declaration to use a package supplying the definitions of the additional macros and environments.

The declaration appears as follows:

```latex
\usepackage{pandoctypes}
```

The package file `pandoctypes.sty` is a component of the Pandoc distribution, empowering any generic LaTeX document, whatever the source, to become a host for document fragments, that is, documents that are not standalone, created in this mode. Users may share this file along with their fragments, to make possible successful inclusion into target host documents.

The package might support options for modifying the names of the additional macros and environments, to improve compatibility in cases of possible name collisions.

### In-line Mode

Under *in-line mode*, generated documents retain the same clear representation as under full mode, but without dependence on definitions given elsewhere. Rather, the generated document body is enclosed in a lexically-scoped group, the contents of which begin with a complete sequence of definitions required for processing of the document contents that follow. The lexical scoping prevents any adverse effect on the processing of the rest of the document.  

Macros are always defined using `\providecommand`, and environments using `\provideenvironment` (available from the package `makecmds`). This constraint has two benefits. First, fragments, that is, documents that are not standalone, are fully safe to include in documents generated under full mode, as well as in documents that otherwise utilize the `pandoctypes` package, or indeed in any other LaTeX documents. Second, definitions given previously in the document for the same names are retained during processing of the fragment, giving the host document the capability to provide any custom definitions as needed.

The basic structure of a fragment generated through in-line mode is as follows:

```latex
\begingroup
% Definitions
\providecommand{\strong}[1]{\textbf{#1}}
% Content
I \strong{very strongly} agree.
\endgroup
```

Although this mode is primarily useful for generating fragments, it may also be employed for generating standalone documents suitable for processing without availability of `pandoctypes.sty`. 

### Verbatim Mode

Under *verbatim mode*, as under in-line mode, generated standalone documents are independent of any external definitions provided in files specific to Pandoc, and moreover, other generated documents are independent of any definitions provided in the host document. Unlike in-line mode, however, only native macros and environments, or those provided by the standard classes and common packages, are invoked in the document body. As such, this mode offers compatibility by essentially emulating previous behavior. In most cases, one of the other modes will provide required compatibility.

## Macros and Environments

The below table summarizes the mapping of abstract document types to any appropriate core LaTeX features, and suggests any further macros and environments needing to be provided by Pandoc for full implementation of the enhancement.

The mapping proposes additional macros or environments in all cases for which no available LaTeX features have the same underlying meaning as some abstract document type, as would derive from a common meaning beyond any use within either Pandoc or LaTeX. Following this principle strictly, the mapping includes additional macros and environments in any case for which none are available whose usage is identical, even if some have usage very similar, to that of an abstract type. This distinction promises clarity and flexibility, without sacrifice of functionality.

---

| Abstract (Pandoc) Type | Available LaTeX Features                                     | Additional Macros and Environments                           |
| ---------------------- | ------------------------------------------------------------ | ------------------------------------------------------------ |
| `CodeBlock`            | `\begin{verbatim}`, `\end{vebatim}`                          |                                                              |
| `BlockQuote`           | `\begin{quote}`, `\end{quote}`                               |                                                              |
| `OrderedList`          |                                                              | `\begin{orderedlist}`, `\end{orderedlist}`, `\item`          |
| `BulletList`           |                                                              | `\begin{bulletlist}`, `\end{bulletlist}`, `\item`            |
| `DefinitionList`       | `\begin{description}`, `\end{description}`, `\item{`*`term`*`}` |                                                              |
| `Header`               | `\chapter{}`, `\section{}`, `\subsection{}`, `\subsubsection{}`, `\paragraph{}`, `\subparagraph{}` |                                                              |
| `HorizontalRule`       |                                                              | `\thematicbreak`                                             |
| `Table`                |                                                              | `\begin{abstracttable}`, `\end{abstracttable}`, `\begin{abstracttablerow},` `\end{abstracttablerow}` |
| `TableCell`            |                                                              | `\begin{abstracttablecell} `, `\end{abstracttablecell}`      |
| `Emph`                 | `\emph{}`                                                    |                                                              |
| `Strong`               |                                                              | `\strong{}`                                                  |
| `Strikeout`            | `\sout{}`                                                    |                                                              |
| `Superscript`          | `\textsuperscript{}`                                         |                                                              |
| `Subscript`            | `\textsubscript{}`                                           |                                                              |
| `Quoted`               | ``` ` ```, `'`, ``` `` ```, `''`                             |                                                              |
| `Cite`                 |                                                              | `\advancedcite[`*`opts`*`]{`*`id`*`}`                        |
| `Code`                 |                                                              | `\inlinecode{}`                                              |
| `Link`                 | `\href{`*`address`*`}{}`                                     |                                                              |
| `Image`                |                                                              | `\image{`*`address`*`}{`*`alttext`*`}`                       |
| `Note`                 | `\footnote{}`                                                |                                                              |

# Discussion

Presently, two methods have been identified, supported by existing application features, for realizing some of the functional benefit of the proposed enhancement. Whereas some users with specific use cases may encounter satisfactory results with either of them, none realizes the promise of an accessible and robust means of full document styling.  

The need for such a solution persists despite the availability of certain methods of approximating its effect.

## Filters

A user might develop a Pandoc filter that replaces each node the writer handles undesirably with a `RawBlock` or `RawInline` node invoking a custom macro or environment. The definitions for all such macros and environments would be placed in the header of each document generated using this filter.

For any except a frequent, enthusiastic user of Pandoc, this method appears prohibitively daunting and taxing, and even a user of this sort could not enjoy any benefit before investing the effort. This user, moreover, would be alone responsible for enhancements, corrections, and other maintenance.

A project maintained by a third party might seem a more appealing prospect. Although such a project may benefit some users, it would incur complexities, as any other project, such as documentation, support, maintenance, and distribution. Challenges would arise from the need to coordinate features of the third-party filter with development of Pandoc. Ultimately, users might become reliant on the uncertain assumption that the project will retain the development resources and interest to remain supported, healthy, and active.

Importantly, any benefit from a third-party solution to any particular user depends on the user's independent discovery, acquisition, learning, and integration of the project.

## Raw Content

Pandoc supports verbatim processing of LaTeX content, through the `raw_tex` extension, embedded in document input of certain formats. Some users may enjoy the benefit of writing LaTeX directly into a source document, but even beyond the restriction on format, this approach carries severe limitations.

Notably, use of raw LaTeX to achieve a target document with full semantic demarcation demands that the document author write LaTeX fragments for all demarcation not handled suitably by Pandoc. This responsibility entails both correct identification of which kinds of demarcation to give in raw representation, and consistent application of this distinction. Consequences of this approach include ongoing obstruction of the mental flow of authoring, and instability of results across output formats.

Indeed, the simplicity of demarcating a document, with attention directed toward production of text, is a core principle of certain formatting languages, and in some cases the cause of their most enthusiastic and consistent support among users. In contrast, intense mixture of two languages greatly undermines the benefit of any single source format. It further undermines the larger benefit of a conversion promising that the essential contents of a document created under the constraints of one format might be transferred to a representation in another format.

# Remarks

## Further Benefits

The proposed enhancement offers a handful of benefits beyond the main objectives.

First, a generated LaTeX document would resemble closely a document that a user might write directly in the LaTeX language. As such, the document could be adopted as a source document, in a case of wanting to migrate, for ongoing revision, text originally represented in another format.

Second, conversion between the LaTeX format and the abstract document model would be nearly completely reversible. Indeed, as all possible semantic document features would be captured in LaTeX source, those features could transfer through Pandoc to another document format. Authors already more comfortable in LaTeX than other formatting languages, as well as those required to write directly in LaTeX for reasons of collaboration, might easily publish in other formats.

## Implementation Notes

Realization of the proposed solution appears to require no major restructuring within the code base. Development of multiple operational modes may entail an expansion of code size. Development of the external LaTeX package would require additional effort and moderate restructuring.  The overall design easily accommodates incremental approach, without affecting release schedule, of the total feature set.

A few further notes are offered concerning implementation.

The first is that the overall objectives may support an evolution toward slightly higher levels of abstraction of macro usage than currently proposed. Such auxiliary improvements have the particular advantage of superior accommodation for differences among document classes, especially the mapping of macros to section levels for books versus articles, while preserving across both cases a uniform representation in LaTeX source of the document body.

The second is that reversible conversions of generated documents might be facilitated by enclosing the document body content in a LaTeX environment, with an empty definition, having the exclusive purpose of allowing the body to be isolated by an automated conversions process.

The third and final note is that the LaTeX writer could share implementation details with a proposed TeX writer. However, output from the LaTeX writer that is restricted to plain TeX offers less benefit to users for a variety of reasons than output that also utilizes LaTeX-specific language features. As such, each of the two writers should maintain a separate layer for resolving this difference.

## Design Limitations

The current proposal is considered comprehensive for the design objectives. No systemic limitations have been identified. The scope, however, is exclusive of support for styling *div* and *span* regions, as would be necessary, and combined with the current solution, sufficient, for entirely general document styling.

**Comment by mb21** [comment]
See extensive discussion in https://github.com/jgm/pandoc/issues/5879

**Comment by tarleb** [comment]
I believe this proposal is good and implicitly included in #1541; closing in favor of that issue.

**Comment by brainchild0** [comment]
Sure, @tarleb, I now see how the proposed would fall inside of #1541, but this particular topic might be a rather comprehensive one, particularly as revealed by earlier discussion, that merits a separate space to resolve various open questions.

Actually, to be completely frank, I often lack a sense of how to  contribute productively  to the conversations for this project. Recall that I earlier posted a feature request (#5879, referenced above) that was much shorter than this posting,  but still  longer than an ordinary issue report. The feedback in that discussion was largely centered on opinions that the idea lacked adequate value or safety, and that the onus rested on the proposal itself to explain the merits and to resolve  (largely unstated, hypothetical) vulnerabilities. That discussion apparently having reached  a dead end, I then  opened this one.

As is plain from reviewing the document,  I took some care to argue the benefits, to anticipate the dangers, and to consider the fine details. While I am not deeply concerned about which issue topics stay open or get closed, much less am I seeking praise, I feel it may be helpful if the proposed design details would receive  consideration, given  the context from which they emerged, even if they might be eventually superseded by ones that are more useful or popular. At least, otherwise i feel confused about what form of posting is actually likely to provoke robust discussion over  how the application features may be usefully refined and expanded.

**Comment by tarleb** [comment]
I sympathize with this and appreciate the frankness. To be quite honest myself: I don't know. We don't have 100% clear cut rules, and personally, I just play things by ear and also wouldn't want to impose additional limitations to the discourse. I closed this issue not because it was not helpful -- very much to the contrary -- but because I'm trying to keep the issue tracker from accumulating "unactionable" items. The future person acting on #1541 will hopefully be thoughtful enough to check the available info, which now includes this issue. It is now linked to from there. I didn't feel competent to summarize the content, or I would have added a brief comment there with a more explicit link to this issue. Please feel free to add such a comment to #1541.

In general, directly actionable items are best kept here in the tracker, with more open discussions better moved to the pandoc-discuss mailing list. Of course, the distinction between actionable and unactionable design discussions is fuzzy, and I've been guilty of breaking this rule myself. So I'm afraid I can't offer a conclusive answer.

**Comment by brainchild0** [comment]
Thanks for sharing thoughts, @tarleb. I have experienced enough of participants stepping on each other's toes to argue strongly in favor of retooling the discussion space and adding at least minimal rules and moderation.

The preferred use of the mailing list versus the issue tracker is a recurrent theme. Submitting a document such as this one to the mailing list would be too difficult, and the effects would be too limited. A web forum may be a more flexible and accessible platform, compared to a mailing list, even a mailing list with a web archive, for support questions and general discussion. For design and feature discussion, I believe that the issue tracker is far superior. I am not fully sure what you intend to mean by "actionable", but tagging an issue according to its purpose, whether making a simple commit or holding a larger discussion, seems more helpful than closing one because it serves a broader purpose than others.

While I believe the issue tracker is a useful place even for general discussions, if a preference prevails for limiting its use to topics of more narrow scope, then I would suggest opening another venue for design and feature discussion. A wiki may serve the needed functions, supporting an internal structure to express a current consensus, with a separate section for "talk". A wiki may also support a space for open contributions of new proposals, which are easy for a visitor to review according to general topic. 

I have posted several ideas that rested on careful consideration, and yet been met with objections seeming to emerge from narrow perspective or incomplete context. While I make no claim to being beyond error or having elevated privilege, I do submit that robustly cited and justified arguments deserve evaluation from the strongest representation of their merits, irrespective of existing preferences. Ideas are not wrong because they challenge consensus, or right because they follow from intuition. Some rules and moderation may have a place, when needed, to protect the tenets of quality discussion, including promoting strong [logical structure](https://en.wikipedia.org/wiki/Deductive_reasoning) and avoiding [straw man arguments](https://en.wikipedia.org/wiki/Straw_man), [arguments from ignorance](https://en.wikipedia.org/wiki/Argument_from_ignorance), and other [fallacious kinds of inference](https://en.wikipedia.org/wiki/Fallacy).

I am glad you like the document, though my only wish is for a quality discussion that leads to inclusion of the greatest feasible level of usability and functionality of the application. I hold enthusiasm for making further contributions, as long as adequate clarity is available for how to provoke a quality conversation.
